### PR TITLE
🎉  Source Facebook Marketing: Added level into custom ads insights

### PR DESCRIPTION
**Pull request by carlonuccio**
## What
By default, we have 'ad' as level for custom ads insights. This PR will enable to choice which level we want data, account/campaign/adset/ad level. By default still remain ad.

## How
In custom ads insights, added a parameter Level.

## Recommended reading order
1. `spec.py`
2. `source.py`
3. `base_insights_streams.py`
4. `test_base_insights_streams.py`

## 🚨 User Impact 🚨
By default, it will have the same behaviour as is.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

```
Name                                                        Stmts   Miss  Cover
-------------------------------------------------------------------------------
source_facebook_marketing/utils.py                             24      0   100%
source_facebook_marketing/streams/common.py                    44      0   100%
source_facebook_marketing/streams/async_job.py                217      0   100%
source_facebook_marketing/streams/__init__.py                   2      0   100%
source_facebook_marketing/spec.py                              42      0   100%
source_facebook_marketing/__init__.py                           2      0   100%
source_facebook_marketing/source.py                            49      1    98%
source_facebook_marketing/streams/async_job_manager.py         78      3    96%
source_facebook_marketing/api.py                               96      9    91%
source_facebook_marketing/streams/base_insight_streams.py     137     14    90%
source_facebook_marketing/streams/base_streams.py             138     30    78%
source_facebook_marketing/streams/streams.py                  118     27    77%
-------------------------------------------------------------------------------
TOTAL                                                         947     84    91%
```
 

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by sathiscode — COMMENTED** [review]
Changes looks good to me.

**Comment by carlonuccio** [comment]
Added Level to spec.json. Let me know if any other code changes are required

**Comment by sathiscode** [comment]
okay, this is what i've so far. The actual spec generated is `"level": {
                "title": "Level",
                "description": "Chosen level for API",
                "default": "ad",
                "allOf": [
                  {
                    "title": "ValidLevel",
                    "description": "Generic enumeration.\n\nDerive from this class to define new enumerations.",
                    "enum": [
                      "account",
                      "ad",
                      "adset",
                      "campaign"
                    ]
                  }
                ]
              },` as per OpenAPI 3 schema. However, the latest source-acceptance-test seems not to handle this correctly.

**Comment by mk-nuccioc** [comment]
what can I do to solve it?

**Comment by carlonuccio** [comment]
what can I do to solve it?

**Comment by sathiscode** [comment]
> what can I do to solve it?

for now, you can change the type of level to string (instead of enum) and update the spec accordingly.

**Comment by carlonuccio** [comment]
edited with array with minItems and maxItems equals to 1. What you think about it?

**Comment by sathiscode** [comment]
Yes, this looks good. Can you please add a TODO comment to revisit and fix as future reference. 
@sajarin Can you please run the acceptance test.

**Comment by carlonuccio** [comment]
I converted again to type=string, introducing pattern, 'cause type array doesn't work. Let me know

**Comment by sathiscode** [comment]
> I converted again to type=string, introducing pattern, 'cause type array doesn't work. Let me know

This is looks better than array option. i'll run the test locally and update

**Comment by sathiscode** [comment]
local test has completed successfully. @sajarin can you please run the acceptance test.

**Comment by sathiscode** [comment]
@carlonuccio test failed as typing.Literal is imported, but not used. Can you please remove Literal import.

image


@sajarin Can you help me with what test to run to capture these errors? I did docker integration test, which didn't report this error.

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by sajarin** [comment]
hey @sathiscode to check for python formatting errors you can run the command below: 

`./gradlew :airbyte-integrations:connectors:source-facebook-marketing:flakeCheck`

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json
Comment (grubberr) [inline airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json:72]: Can we use ?

```json
"type" : "string",
"enum": ["ad", "adset", "campaign", "account"]
```

**Review by grubberr — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json
Comment (carlonuccio) [reply airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json:72]: on first commit 063c59056d281170183aeaa5e313fdabf8ac2b38, I tried

`
ValidLevel = Enum("ValidLevel", AdsInsights.Level.__dict__)

level: Optional[ValidLevel] = Field(
    title="Level",
    description="Chosen level for API",
    default="ad",
)
`

But integration tests didn't work with this

**Comment by carlonuccio** [comment]
Do you have any updates?

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json
Comment (grubberr) [reply airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json:72]: @carlonuccio can you re-implement using enum? and I will  help with tests

Comment (carlonuccio) [reply airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json]: edited, let me know

**Comment by sajarin** [comment]
@grubberr are we okay to merge this?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py
Comment (grubberr) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:35]: Can we remove this?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py
Comment (grubberr) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:54]: Do we really need this?

**Comment by sajarin** [comment]
Thanks @carlonuccio for addressing the changes. @grubberr could I get your approval here if you feel this PR is a good place?

**Review by grubberr — APPROVED** [review]

**Comment by mk-nuccioc** [comment]
We can't merge until https://github.com/airbytehq/airbyte/pull/19151 is ready, right?

**Comment by carlonuccio** [comment]
We can't merge until https://github.com/airbytehq/airbyte/pull/19151 is ready, right?

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by carlonuccio** [comment]
Hi @sathiscode @grubberr, can we consider to merge this?

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**
